src/Build/BackEnd/Client/MSBuildClient.cs

```diff
@@ -265,9 +265,10 @@ private bool TryShutdownServer(CancellationToken cancellationToken)
                 return true;
             }
 
-            // Check that server is not busy.
-            bool serverWasBusy = ServerWasBusy();
-            if (serverWasBusy)
+            // Check and wait for server to be not busy for some short time to avoid race condition when server reports build is finished but had not released ServerBusy mutex yet.
```

Comment:
Would it make sense to ensure this never happens by releasing ServerBusy before reporting the finish? Or does that violate other invariants?

Reply:
I thought about it. It is hard to me to identify and understand all consequences. In theory it shall be fine, report server not busy when its namedpipe connection still sending data to finishing client shall be fine, since worst case we fall back to non server because namedpipe could not be opened. But is sending final build result to client part of being busy? I think it is? So declaring not busy while still doing something related with just finished build is a little bit laying :-)

All in all, I just consider my changes slightly less risky.